src/dbnode/storage/index/block.go

```diff
@@ -711,27 +715,58 @@ func (b *block) aggregateWithSpan(
 			}
 
 			field, term := iter.Current()
-			batch, numAdded = b.appendFieldAndTermToBatch(batch, field, term, iterateTerms)
-			currBatchSize += numAdded
 
-			// continue appending to the batch until we hit our max batch size
-			if currBatchSize < maxBatch {
-				continue
-			}
-
-			// update recently queried docs to monitor memory.
+			// TODO: remove this legacy doc tracking implementation when alternative
+			// limits are in place.
 			if results.EnforceLimits() {
-				if err := b.docsLimit.Inc(len(batch), source); err != nil {
-					return false, err
+				if lastField == nil {
```

Comment:
so we're essentially only incrementing the docs limit if we're working on a new field?

Reply:
So essentially what was happening previously was this:

- incoming field/term tuples were added to the batch [here]
-- if terms were included in the query (e.g. `label/foo/values`), and the term matched the previous term, it would just alloc and append the field onto the existing entry’s terms; this was the underlying problem that took out our test cluster when their cardinality exploded, as it would allocate a huge slice of tags, as well as the tags themselves, here. This is because the batch size is calculated as len(terms), and would cause the batch max to apply only to new(different) label names.
-- Once either batch is full (note that for a label/foo/values request, this would never happen since the batch length would be at most 1) or incoming entries are exhausted, we’d increment the docs limit with len(terms).

Although there were no changes to how the docs limit was incremented, because the batch size calculation was changed from len(terms) to  ~~`len(terms) + len(all entries)`, we’d increment the docs limit far more often, which is a big problem when the terms are mostly a single value, as is always the case with `label/foo/values`, which is probably the most widely used metadata query endpoint; in essence we'd be reporting a doc increase for each 256 terms, rather than once per query as previously.